File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <stack>
#include <math.h>

using namespace std;

vector<bool> v, vw;
vector<vector<int>> w;
int sz;

int dfs(){
    stack<int> st;
    st.push(1);
    v[1] = 1;
    
    while(!st.empty()){
        int x = st.top(); st.pop();
        for(int i = 0; i < sz - 1; i++){
            if(vw[i]) continue;
            int a = w[i][0], b = w[i][1];
            if(!v[a] && v[b]){
                vw[i] = v[a] = 1;
                st.push(a);
                break;
            } 
            if(v[a] && !v[b]){
                vw[i] = v[b] = 1;
                st.push(b);
                break;
            }
        }   
    }
    int cnt = 0;
    for(int i = 1; i <= sz; i++) if(v[i]) cnt++;
    return cnt;
}

int solution(int n, vector<vector<int>> wires) {
    int ans = n;
    w = wires;
    sz = n;
    
    vw.resize(n, 0); v.resize(n + 1, 0);
    
    for(int i = 0; i < n - 1; i++){
        fill(v.begin(), v.end(), 0); fill(vw.begin(), vw.end(), 0);
        vw[i] = 1;
        int tmp = abs(n - 2 * dfs());
        if(ans > tmp) ans = tmp;
    }
    return ans;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    boolean[] visit = null;
    Map<Integer, List<Integer>> m = new HashMap<>();
   
    
    public int solution(int n, int[][] wires) {
        int answer = n, start = 0;
        visit = new boolean[n];
        
        for(int i = 0; i < wires.length; i++){
            m.put(wires[i][0], m.getOrDefault(wires[i][0], new ArrayList<>()));
            m.get(wires[i][0]).add(wires[i][1]);
            m.put(wires[i][1], m.getOrDefault(wires[i][1], new ArrayList<>()));
            m.get(wires[i][1]).add(wires[i][0]);
        }
     
        for(int i = 0; i < wires.length; i++) {
            Arrays.fill(visit, false);
            
            if(i == 0) start = wires[1][0];
            else start = wires[i - 1][0];
            
            m.get(wires[i][0]).remove(Integer.valueOf(wires[i][1]));
            m.get(wires[i][1]).remove(Integer.valueOf(wires[i][0]));
            
           
            bfs(start);
            int cnt = 0;
            for(int j = 0; j < n; j++) if(visit[j]) cnt++;
            
            answer = Math.min(answer, Math.abs(n - cnt - cnt));
            
            m.get(wires[i][0]).add(Integer.valueOf(wires[i][1]));
            m.get(wires[i][1]).add(Integer.valueOf(wires[i][0]));
        }
        
        return answer;
    }
    
    public void bfs(int start){
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        visit[start - 1] = true;

        while(!q.isEmpty()){
            int cur = q.poll();
            List<Integer> tmp = m.get(cur);

            for(Integer nxt : tmp) {
                if(visit[nxt - 1]) continue;
                visit[nxt - 1] = true;
                q.add(nxt);
            }
        }
    }
}
